File: sqlparser/src/sql/csgen/SqlUtilities.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Net;
using System.Net.Sockets;
using System.Text;
using RSSBus.core.j2cs;




using RSSBus.core;
using RSSBus;
using System.Collections.Generic;
namespace CData.Sql {

public sealed class SqlUtilities {
  public const string CACHE_SUFFIX = "#Cache";
  public const string DELETED_SUFFIX = "#Deleted";
  public const string TEMP_SUFFIX = "#Temp";

  public const string CDATA_SQL_VARCHAR = "VARCHAR";
  public const string CDATA_SQL_BIT = "BIT";
  public const string CDATA_SQL_BINARY = "BINARY";
  public const string CDATA_SQL_BIGINT = "BIGINT";
  public const string CDATA_SQL_INTEGER = "INTEGER";
  public const string CDATA_SQL_SMALLINT = "SMALLINT";
  public const string CDATA_SQL_FLOAT = "FLOAT";
  public const string CDATA_SQL_DOUBLE = "DOUBLE";
  public const string CDATA_SQL_DECIMAL = "DECIMAL";
  public const string CDATA_SQL_DATE = "DATE";
  public const string CDATA_SQL_TIME = "TIME";
  public const string CDATA_SQL_TIMESTAMP = "TIMESTAMP";

  public static string GetSqlType(int typeCode) {
    if (typeCode == ColumnInfo.DATA_TYPE_BINARY) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_VARBINARY) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BLOB) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BOOLEAN) {
      return CDATA_SQL_BIT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DATE) {
      return CDATA_SQL_DATE;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TIMESTAMP) {
      return CDATA_SQL_TIMESTAMP;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TIME) {
      return CDATA_SQL_TIME;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DECIMAL) {
      return CDATA_SQL_DECIMAL;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DOUBLE) {
      return CDATA_SQL_DOUBLE;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TINYINT) {
      return CDATA_SQL_SMALLINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_SMALLINT) {
      return CDATA_SQL_SMALLINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_INTEGER) {
      return CDATA_SQL_INTEGER;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BIGINT) {
      return CDATA_SQL_BIGINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_FLOAT) {
      return CDATA_SQL_FLOAT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_VARCHAR) {
      return CDATA_SQL_VARCHAR;
    } else {
      return CDATA_SQL_VARCHAR;
    }
  }
  
  public static bool IsCacheTable(string tableName) {
    return tableName.ToLower().EndsWith(CACHE_SUFFIX.ToLower());
  }

  public static bool IsTempTable(string tableName) {
    return tableName.ToLower().EndsWith(TEMP_SUFFIX.ToLower());
  }

  public static bool IsDeletedTable(string tableName){
    return tableName.ToLower().EndsWith(DELETED_SUFFIX.ToLower());
  }

  public static bool IsKnownAggragation(string formulaName) {
    bool isAggragation = false;
    string [] knownAggragation = new string [] {"AVG", "SUM", "MIN", "MAX", "COUNT", "STDEV", "STDEVP", "VAR", "VARP", "COUNT_BIG"};
    if (formulaName != null) {
      foreach(String name in knownAggragation) {
        if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(name, formulaName)) {
          isAggragation = true;
          break;
        }
      }
    }
    return isAggragation;
  }

  public static bool IsSimpleTable(SqlTable table) {
    return !table.HasJoin() && !table.IsNestedQueryTable() && !table.IsNestedJoinTable() && !table.IsFunctionValueTable();
  }

  public static SqlCollection<SqlTable> GetSourceTables(SqlStatement stmt) {
    SqlCollection<SqlTable> sources = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(sources, new SourceTableMatcher());
    stmt.Accept(collector);
    return sources;
  }

  public static SqlCollection<SqlTable> GetTables(SqlStatement stmt, ITableMatch matcher) {
    SqlCollection<SqlTable> sources = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(sources, matcher);
    stmt.Accept(collector);
    return sources;
  }

  public static   void FetchElement<T> (SqlCollection<T> inSets, SqlCollection<T> outSets, ISqlElementComparable<T> comparable)  where T : ISqlElement {
    foreach(T t in inSets) {
      if (0 == comparable.CompareTo(t)) {
        outSets.Add(t);
      }
    }
  }

  public static void CheckQuerySemantic(SqlQueryStatement queryStatement) {
    if (queryStatement.GetTable() == null) return;

    SqlCollection<SqlTable> references = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(references, new TableAliasMatcher());
    queryStatement.Accept(collector);
    queryStatement.Accept(new SqlQueryAliasChecker(references));

    queryStatement.Accept(new SqlColumnSourceChecker());
  }

  public static SqlTable BuildTempTable(SqlTable sqlTable) {
    return new SqlTable(sqlTable.GetCatalog(), sqlTable.GetSchema(), sqlTable.GetName() + TEMP_SUFFIX);
  }

  public static void CollectParameters(SqlConditionNode criteria, IList<QueryParameter> collector) {
    if (criteria == null || collector == null) {
      return;
    }

    if (criteria is SqlCriteria) {
      SqlCriteria c = (SqlCriteria) criteria;
      if (c.IsRightParameter()) {
        //TODO: Should we support const, function and more?
        if (c.GetLeft() is SqlGeneralColumn) {
          collector.Add(new QueryParameter(((SqlGeneralColumn) c.GetLeft()).GetColumnName(), null));
        }
      } else if (c.IsLeftParameter()) {
        if (c.GetRight() is SqlGeneralColumn) {
          collector.Add(new QueryParameter(((SqlGeneralColumn) c.GetRight()).GetColumnName(), null));
        }
      }
    } else if (criteria is SqlCondition) {
      SqlCondition c = (SqlCondition)criteria;
      if (c.GetLeft() is SqlConditionNode) {
        CollectParameters((SqlConditionNode)c.GetLeft(), collector);
      }
      if (c.GetRight() is SqlConditionNode) {
        CollectParameters((SqlConditionNode)c.GetRight(), collector);
      }
    } else if (criteria is SqlConditionNot) {
      SqlExpression c = ((SqlConditionNot) criteria).GetCondition();
      if (c is SqlConditionNode) {
        CollectParameters((SqlConditionNode) c, collector);
      }
    }
  }
}

sealed class SourceTableMatcher : ITableMatch {

  public SqlTable Create(SqlTable table) {
    return new SqlTable(table.GetCatalog(),
            table.GetSchema(),
            table.GetName());
  }

  public bool Accept(SqlTable t, TablePartType type) {
    return (TablePartType.SimpleTable  == type
            && ParserCore.IsValidTableName(t.GetName()));
  }

  public bool Unwind(SqlTable t, TablePartType type) {
    return TablePartType.NestedQuery == type
        || TablePartType.NestedJoin == type;
  }
}

sealed class TableAliasMatcher : ITableMatch {

  public SqlTable Create(SqlTable table) {
    return new SqlTable(table.GetCatalog(),
            table.GetSchema(),
            table.GetName(),
            table.GetAlias());
  }

  public bool Accept(SqlTable t, TablePartType type) {
    return true;
  }

  public bool Unwind(SqlTable t, TablePartType type) {
    return true;
  }
}

sealed class TableOnlyMatcher : ITableMatch {

  public SqlTable Create(SqlTable table) {
    return new SqlTable(table.GetCatalog(),
            table.GetSchema(),
            table.GetName(),
            table.GetAlias());
  }

  public bool Accept(SqlTable t, TablePartType type) {
    return type != TablePartType.JoinPart;
  }

  public bool Unwind(SqlTable t, TablePartType type) {
    return true;
  }
}
}

File: sqlparser/src/sql/SqlUtilities.java
//@
package cdata.sql;
import core.ParserCore;

import java.util.List;
//@

/*#
using RSSBus.core;
using RSSBus;
using System.Collections.Generic;
namespace CData.Sql {
#*/
public final class SqlUtilities {
  public static final String CACHE_SUFFIX = "#Cache";
  public static final String DELETED_SUFFIX = "#Deleted";
  public static final String TEMP_SUFFIX = "#Temp";

  public static final String CDATA_SQL_VARCHAR = "VARCHAR";
  public static final String CDATA_SQL_BIT = "BIT";
  public static final String CDATA_SQL_BINARY = "BINARY";
  public static final String CDATA_SQL_BIGINT = "BIGINT";
  public static final String CDATA_SQL_INTEGER = "INTEGER";
  public static final String CDATA_SQL_SMALLINT = "SMALLINT";
  public static final String CDATA_SQL_FLOAT = "FLOAT";
  public static final String CDATA_SQL_DOUBLE = "DOUBLE";
  public static final String CDATA_SQL_DECIMAL = "DECIMAL";
  public static final String CDATA_SQL_DATE = "DATE";
  public static final String CDATA_SQL_TIME = "TIME";
  public static final String CDATA_SQL_TIMESTAMP = "TIMESTAMP";

  public static String getSqlType(int typeCode) {
    if (typeCode == ColumnInfo.DATA_TYPE_BINARY) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_VARBINARY) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BLOB) {
      return CDATA_SQL_BINARY;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BOOLEAN) {
      return CDATA_SQL_BIT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DATE) {
      return CDATA_SQL_DATE;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TIMESTAMP) {
      return CDATA_SQL_TIMESTAMP;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TIME) {
      return CDATA_SQL_TIME;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DECIMAL) {
      return CDATA_SQL_DECIMAL;
    } else if (typeCode == ColumnInfo.DATA_TYPE_DOUBLE) {
      return CDATA_SQL_DOUBLE;
    } else if (typeCode == ColumnInfo.DATA_TYPE_TINYINT) {
      return CDATA_SQL_SMALLINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_SMALLINT) {
      return CDATA_SQL_SMALLINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_INTEGER) {
      return CDATA_SQL_INTEGER;
    } else if (typeCode == ColumnInfo.DATA_TYPE_BIGINT) {
      return CDATA_SQL_BIGINT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_FLOAT) {
      return CDATA_SQL_FLOAT;
    } else if (typeCode == ColumnInfo.DATA_TYPE_VARCHAR) {
      return CDATA_SQL_VARCHAR;
    } else {
      return CDATA_SQL_VARCHAR;
    }
  }
  
  public static boolean isCacheTable(String tableName) {
    return tableName.toLowerCase().endsWith(CACHE_SUFFIX.toLowerCase());
  }

  public static boolean isTempTable(String tableName) {
    return tableName.toLowerCase().endsWith(TEMP_SUFFIX.toLowerCase());
  }

  public static boolean isDeletedTable(String tableName){
    return tableName.toLowerCase().endsWith(DELETED_SUFFIX.toLowerCase());
  }

  public static boolean isKnownAggragation(String formulaName) {
    boolean isAggragation = false;
    String [] knownAggragation = new String [] {"AVG", "SUM", "MIN", "MAX", "COUNT", "STDEV", "STDEVP", "VAR", "VARP", "COUNT_BIG"};
    if (formulaName != null) {
      for (String name : knownAggragation) {
        if (name.equalsIgnoreCase(formulaName)) {
          isAggragation = true;
          break;
        }
      }
    }
    return isAggragation;
  }

  public static boolean isSimpleTable(SqlTable table) {
    return !table.hasJoin() && !table.isNestedQueryTable() && !table.isNestedJoinTable() && !table.isFunctionValueTable();
  }

  public static SqlCollection<SqlTable> getSourceTables(SqlStatement stmt) throws Exception {
    SqlCollection<SqlTable> sources = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(sources, new SourceTableMatcher());
    stmt.accept(collector);
    return sources;
  }

  public static SqlCollection<SqlTable> getTables(SqlStatement stmt, ITableMatch matcher) throws Exception {
    SqlCollection<SqlTable> sources = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(sources, matcher);
    stmt.accept(collector);
    return sources;
  }

  public static /*@*/<T extends ISqlElement> void fetchElement/*@*/ /*# void FetchElement<T>#*/ (SqlCollection<T> inSets, SqlCollection<T> outSets, ISqlElementComparable<T> comparable) /*# where T : ISqlElement #*/{
    for (T t : inSets) {
      if (0 == comparable.compareTo(t)) {
        outSets.add(t);
      }
    }
  }

  public static void checkQuerySemantic(SqlQueryStatement queryStatement) throws Exception {
    if (queryStatement.getTable() == null) return;

    SqlCollection<SqlTable> references = new SqlCollection<SqlTable>();
    TableFinder collector = new TableFinder(references, new TableAliasMatcher());
    queryStatement.accept(collector);
    queryStatement.accept(new SqlQueryAliasChecker(references));

    queryStatement.accept(new SqlColumnSourceChecker());
  }

  public static SqlTable buildTempTable(SqlTable sqlTable) throws Exception {
    return new SqlTable(sqlTable.getCatalog(), sqlTable.getSchema(), sqlTable.getName() + TEMP_SUFFIX);
  }

  public static void collectParameters(SqlConditionNode criteria, List<QueryParameter> collector) {
    if (criteria == null || collector == null) {
      return;
    }

    if (criteria instanceof SqlCriteria) {
      SqlCriteria c = (SqlCriteria) criteria;
      if (c.isRightParameter()) {
        //TODO: Should we support const, function and more?
        if (c.getLeft() instanceof SqlGeneralColumn) {
          collector.add(new QueryParameter(((SqlGeneralColumn) c.getLeft()).getColumnName(), null));
        }
      } else if (c.isLeftParameter()) {
        if (c.getRight() instanceof SqlGeneralColumn) {
          collector.add(new QueryParameter(((SqlGeneralColumn) c.getRight()).getColumnName(), null));
        }
      }
    } else if (criteria instanceof SqlCondition) {
      SqlCondition c = (SqlCondition)criteria;
      if (c.getLeft() instanceof SqlConditionNode) {
        collectParameters((SqlConditionNode)c.getLeft(), collector);
      }
      if (c.getRight() instanceof SqlConditionNode) {
        collectParameters((SqlConditionNode)c.getRight(), collector);
      }
    } else if (criteria instanceof SqlConditionNot) {
      SqlExpression c = ((SqlConditionNot) criteria).getCondition();
      if (c instanceof SqlConditionNode) {
        collectParameters((SqlConditionNode) c, collector);
      }
    }
  }
}

final class SourceTableMatcher implements ITableMatch {

  public SqlTable create(SqlTable table) {
    return new SqlTable(table.getCatalog(),
            table.getSchema(),
            table.getName());
  }

  public boolean accept(SqlTable t, TablePartType type) {
    return (TablePartType.SimpleTable  == type
            && ParserCore.isValidTableName(t.getName()));
  }

  public boolean unwind(SqlTable t, TablePartType type) {
    return TablePartType.NestedQuery == type
        || TablePartType.NestedJoin == type;
  }
}

final class TableAliasMatcher implements ITableMatch {

  public SqlTable create(SqlTable table) {
    return new SqlTable(table.getCatalog(),
            table.getSchema(),
            table.getName(),
            table.getAlias());
  }

  public boolean accept(SqlTable t, TablePartType type) {
    return true;
  }

  public boolean unwind(SqlTable t, TablePartType type) {
    return true;
  }
}

final class TableOnlyMatcher implements ITableMatch {

  public SqlTable create(SqlTable table) {
    return new SqlTable(table.getCatalog(),
            table.getSchema(),
            table.getName(),
            table.getAlias());
  }

  public boolean accept(SqlTable t, TablePartType type) {
    return type != TablePartType.JoinPart;
  }

  public boolean unwind(SqlTable t, TablePartType type) {
    return true;
  }
}

